libr/bin/format/elf/elf.c

```diff
@@ -488,12 +488,40 @@ ut64 Elf_(r_bin_elf_get_main_offset)(struct Elf_(r_bin_elf_obj_t) *bin) {
 
 	// MIPS
 	/* get .got, calculate offset of main symbol */
-	if (!memcmp (buf, "\x21\x00\xe0\x03\x01\x00\x11\x04\x00\x00\x00\x00", 12)) {
-		ut64 got_addr = 0LL; // TODO: get .got offset
-		short delta = (buf[28]+(buf[29]<<8));
-		// NOTE: This is the way to resolve 'gp' register
-		r_buf_read_at (bin->b, got_addr+(32734+delta), buf, 4);
-		return (ut64)((int)(buf[0]+(buf[1]<<8)+(buf[2]<<16)+(buf[3]<<24)))-bin->baddr;
+	if (!memcmp (buf, "\x21\x00\xe0\x03\x01\x00\x11\x04", 8)) {
+
+		/*
+		    assuming the startup code looks like
+		        got = gp-0x7ff0
+		        got[index__libc_start_main] ( got[index_main] );
+
+		    looking for the instruction generating the first argument to find main
+		        lw a0, offset(gp)
+		*/
+
+		ut64 got_offset;
+
+		if ((got_offset = Elf_(r_bin_elf_get_section_offset) (bin, ".got")) == -1 &&
+		    (got_offset = Elf_(r_bin_elf_get_section_offset) (bin, ".got.plt")) == -1)
+		{
+			return 0;
+		}
+
+		ut64 gp = got_offset + 0x7ff0;
+
+		unsigned i;
```

Comment:
Dont use undefined types. Plain int should be fine. But you can use ut32 if really 2GB Limit matters